chairman/Chairman.hs

```diff
@@ -0,0 +1,434 @@
+-- | Rollback a single block.  If the rollback point is not found, we simply
+-- error.  It should never happen if the security parameter is set up correctly.
+--
+rollback
+    :: forall blk m.
+       ( MonadSTM m
+       , HasHeader blk
+       )
+    => CoreNodeId
+    -> ChainsVar m blk
+    -> Point blk
+    -> STM m ()
+rollback coreNodeId chainsVar p =
+    modifyTVar' chainsVar (Map.adjust fn coreNodeId)
+  where
+    fn :: ChainFragment blk -> ChainFragment blk
+    fn cf = case CF.rollback p cf of
+      Nothing  -> CF.Empty
+      Just cf' -> cf'
+
+
+--
+-- Client Application
+--
+
+localInitiatorNetworkApplication
+  :: forall blk m peer.
+     ( RunNode blk
+     , Condense blk
+     , MonadAsync m
+     , MonadSTM   m
+     , MonadST    m
+     , MonadThrow m
+     , MonadTimer m
+     )
+  => CoreNodeId
+  -> ChainsVar m blk
+  -> Int
+  -> Maybe BlockNo
+  -> Tracer m (ChairmanTrace blk)
+  -> Tracer m (TraceSendRecv (ChainSync blk (Point blk)) peer DeserialiseFailure)
+  -- ^ tracer which logs all chain-sync messages send and received by the client
+  -- (see 'Ouroboros.Network.Protocol.ChainSync.Type' in 'ouroboros-network'
+  -- package)
+  -> Tracer m (TraceSendRecv (LocalTxSubmission (GenTx blk) String) peer DeserialiseFailure)
+  -- ^ tracer which logs all local tx submission protocol messages send and
+  -- received by the client (see 'Ouroboros.Network.Protocol.LocalTxSubmission.Type'
+  -- in 'ouroboros-network' package).
+  -> NodeConfig (BlockProtocol blk)
+  -> Versions NodeToClientVersion DictVersion
+              (OuroborosApplication InitiatorApp peer NodeToClientProtocols
+                                    m ByteString () Void)
+localInitiatorNetworkApplication coreNodeId chainsVar securityParam maxBlockNo chairmanTracer chainSyncTracer localTxSubmissionTracer pInfoConfig =
+    simpleSingletonVersions
+      NodeToClientV_1
+      (NodeToClientVersionData { networkMagic = 0 })
+      (DictVersion nodeToClientCodecCBORTerm)
+
+  $ OuroborosInitiatorApplication $ \peer ptcl -> case ptcl of
+      LocalTxSubmissionPtcl -> \channel -> do
+        txv <- newEmptyTMVarM @_ @(GenTx blk)
+        runPeer
+          localTxSubmissionTracer
+          localTxSubmissionCodec
+          peer
+          channel
+          (localTxSubmissionClientPeer
+              (txSubmissionClient @(GenTx blk) txv))
+
+      ChainSyncWithBlocksPtcl -> \channel ->
+        runPeer
+          chainSyncTracer
+          (localChainSyncCodec pInfoConfig)
+          peer
+          channel
+          (chainSyncClientPeer $ chainSyncClient chairmanTracer coreNodeId chainsVar securityParam maxBlockNo)
+
+
+--
+-- Submission client
+--
+
+-- | A 'LocalTxSubmissionClient' that submits transactions reading them from
+-- a 'TMVar'.  A real implementation should use a better synchronisation
+-- primitive.  This demo creates and empty 'TMVar' in
+-- 'muxLocalInitiatorNetworkApplication' above and never fills it with a tx.
+--
+txSubmissionClient
```

Comment:
Because local client needs both chain sync and tx submission. It's a limitation of the `ouroboros-network` interface, and partially a limitation of the multiplexer (if any mini-protocol exists, the multiplexer will kill all other mini-protocol threads as well).

Reply:
I see, that's unfortunate.